contractsSol5/KyberNetworkConnector.sol

```diff
@@ -0,0 +1,108 @@
+pragma solidity 0.5.11;
+
+import "./utils/Withdrawable3.sol";
+import "./utils/Utils4.sol";
+import "./utils/zeppelin/ReentrancyGuard.sol";
+import "./utils/zeppelin/SafeERC20.sol";
+import "./IKyberNetwork.sol";
+
+
+/*
+ *   @title Kyber Network Connector main contract
+ *   Interacts with contracts:
+ *       KyberNetwork: to call  fee data
+ *       KyberNetworkProxy: only proxy can call tradeWithHint
+ *
+ *    Kyber Network Connector will call KyberNetwork for:
+ *       - getExpectedRateWithHintAndFee
+ *       - tradeWithHintAndFee
+ */
+
+contract KyberNetworkConnector is
+    Withdrawable3,
+    Utils4,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    uint  internal constant PERM_HINT_GET_RATE = 1 << 255;
+    address internal constant ZERO_ADDRESS = address(uint160(0));
+    IKyberNetwork public kyberNetwork;
+    address public networkProxy;
+
+    constructor(address _admin) public Withdrawable3(_admin) {}
+
+    event KyberProxyAdded(address proxy);
+    event KyberNetworkAdded(address network);
+
+    function addKyberNetwork(address _kyberNetwork) external {
+        onlyAdmin();
+        require(_kyberNetwork != address(0), "proxy 0");
+        kyberNetwork = IKyberNetwork(_kyberNetwork);
+
+        emit KyberNetworkAdded(_kyberNetwork);
+    }
+
+    function addKyberProxy(address _networkProxy) external {
+        onlyAdmin();
+        require(networkProxy != address(0), "proxy 0");
+        networkProxy = _networkProxy;
+
+        emit KyberProxyAdded(networkProxy);
+    }
+
+    function maxGasPrice() public view returns(uint) {
+        return kyberNetwork.maxGasPrice();
+    }
+
+    function enabled() public view returns(bool) {
+        return kyberNetwork.enabled();
+    }
+
+    function getExpectedRate(ERC20 src, ERC20 dest, uint256 srcQty)
+        public
+        view
+        returns (uint256 expectedRate, uint256 slippageRate)
+    {
+        bytes memory hint;
+        ( , expectedRate, ) = kyberNetwork.getExpectedRateWithHintAndFee(src, dest, srcQty, 0, hint);
+        // use simple backward compatible optoin.
+        slippageRate = expectedRate * 97 / 100;
+    }
+
+    function prepareTrade(IERC20 src, uint srcAmount) internal
+    {
+        require(msg.sender == networkProxy, "bad sender");
+        if (src == ETH_TOKEN_ADDRESS) {
```

Comment:
this validation already done on network level.
Don't have to do it here.

its double logic.

so please write connector as small as possible. no extra logic.